azure-iot-device/azure/iot/device/provisioning/pipeline/pipeline_stages_provisioning.py

```diff
@@ -48,3 +63,331 @@ def _execute_op(self, op):
+                )
+
+                # This could be an error that has been reported to this stage Or this
+                # could be an error because the service responded with status code 300
+                error = self._process_error(provisioning_op, REGISTER, error=error)
+
+                if not error:
+                    success_status_code = provisioning_op.status_code
+                    retry_interval = (
+                        int(provisioning_op.retry_after, 10)
+                        if provisioning_op.retry_after is not None
+                        else constant.DEFAULT_POLLING_INTERVAL
+                    )
+                    decoded_response = self._decode_response(provisioning_op)
+                    operation_id = self._get_operation_id(decoded_response)
+
+                    # retry after scenario
+                    if success_status_code >= 429:
+
+                        self_weakref = weakref.ref(self)
+
+                        @pipeline_thread.invoke_on_pipeline_thread_nowait
+                        def do_retry_after():
+                            this = self_weakref()
+                            logger.info(
+                                "{stage_name}({op_name}): retrying".format(
+                                    stage_name=this.name, op_name=op.name
+                                )
+                            )
+                            op.retry_after_timer.cancel()
+                            op.completed = False
+                            this._execute_op(op)
+
+                        logger.warning(
+                            "{stage_name}({op_name}): Op needs retry with interval {interval} because of {error}.  Setting timer.".format(
+                                stage_name=self.name,
+                                op_name=op.name,
+                                interval=retry_interval,
+                                error=error,
+                            )
+                        )
+
+                        op.retry_after_timer = Timer(retry_interval, do_retry_after)
+                        op.retry_after_timer.start()
+
+                    # Service success scenario
+                    else:
+                        registration_status = self._get_registration_status(decoded_response)
+                        if registration_status == "assigned" or registration_status == "failed":
+                            # process complete response here
+                            complete_registration_result = self._form_complete_result(
+                                request_id=provisioning_op.request_id,
+                                operation_id=operation_id,
+                                decoded_response=decoded_response,
+                                status=registration_status,
+                            )
+                            op.registration_result = complete_registration_result
+                            self.complete_op(op, error=error)
```

Comment:
>self.complete_op(op, error=error [](start = 28, length = 32)

At this point, if registration_status == 'failed', it will complete the op with error=None.  I would assume that we would want to complete this op with an error since there was a failure.

Reply:
same comment as below

---
In reply to: [345909570](https://github.com/Azure/azure-iot-sdk-python/pull/391#discussion_r345909570) [](ancestors = 345909570)